Expose the content type associated with roles for ActivityStream objects in the API

jbradberry opened the pull request:
##### SUMMARY
Due to the way ActivityStream objects are created based on actions, an association of two objects via a role could wind up in different order, depending on the direction in which the role grant was performed.  We need an extra piece of information to determine which object is the role's parent.  This PR exposes the content type from the role by manipulating the value constructed in `ActivityStream.object_relationship_type`, which should be sufficient to determine which of the two objects owns it.  We'll need this to arrange the object order in the UI for ActivityStreams associating two objects via roles.

related #3841

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
 - API
 - UI

##### AWX VERSION
```
awx: 4.0.0
```

jbradberry committed:
Expose the role's content_type field and reorder the objects in the UI

when the resource was stored on ActivityStream as object1 instead of object2.

related #3841

jbradberry commented:
After change, the creation:

![Screenshot_2019-05-10 Ansible AWX ACTIVITY STREAM](https://user-images.githubusercontent.com/685957/57548547-374d9800-732f-11e9-8cd1-c30f2bb8ca3b.png)

and granting:

![Screenshot_2019-05-10 Ansible AWX ACTIVITY STREAM(1)](https://user-images.githubusercontent.com/685957/57548578-4f251c00-732f-11e9-9c87-b625b74332ae.png)

AlanCoding commented on awx/api/serializers.py:
```diff
@@ -128,7 +128,7 @@
     'custom_inventory_script': DEFAULT_SUMMARY_FIELDS,
     'source_script': ('name', 'description'),
-    'role': ('id', 'role_field'),
+    'role': ('id', 'role_field', 'content_type'),
```
This could have a fairly significant effect on the number of queries needed to generate list views. However, I don't actually know what views make use of the default role summary fields.

jbradberry replied:
Good point, though Django does cache the full set of `ContentType` objects, so there shouldn't actually be any db hits.

AlanCoding replied:
I thought that the cache only worked if you used the default content type manager, and using it as a ForeignKey or in a GenericForeignKey like here, it'll still go fetch it for each. I agree that having the `content_type_id` field from a role object would allow it to be looked up straight from the cache, but I don't think it'll do that unless you did something like `ContentType.objects. get_for_id(role.content_type_id)`

AlanCoding replied:
but again, for this to ever affect anything at all, some model would have to have a field named `role`, and I don't think this is particularly common.

jbradberry replied:
Yeah, I (after a bit of fumbling) tested the queries involved and it _does_ hit the db.

```
In [34]: context.initial_queries, context.final_queries
Out[34]: (42, 2980)
```

```
In [33]: for x in context.captured_queries:
    ...:     if 'content_type' in x['sql']:
    ...:         print(x['sql'])
(...)
SELECT "django_content_type"."id", "django_content_type"."app_label", "django_content_type"."model" FROM "django_content_type" WHERE "django_content_type"."id" = 20
```

jbradberry replied:
Even if I rearrange things, it's going to be doing an N+1 with regards to the Role objects, so this is a dead end.  I'm going to rethink this a bit.

jbradberry committed:
Instead of exposing Role.content_type, create a new serializer field

called `object_type`, which is constructed based on manipulating the
string value of ActivityStream.object_relationship_type.  Since that
field does have the full class name, this manipulation should match
the manipulation that is done to construct the values of object1 and
object2 when ActivityStream is created.

AlanCoding commented on awx/api/serializers.py:
```diff
@@ -4956,6 +4956,7 @@ class ActivityStreamSerializer(BaseSerializer):
 
     changes = serializers.SerializerMethodField()
     object_association = serializers.SerializerMethodField()
+    object_type = serializers.SerializerMethodField()
```
help_text?

AlanCoding reviewed (changes requested):
This makes a lot of sense to me - having 2 fields to deconstruct the obfuscated internal field object_relationship_type for the client. My final request is that we add help_text to both of these method fields to establish documentation for them.

jbradberry committed:
Add help text to the ActivityStreamSerializer m2m relationship fields

AlanCoding commented on awx/api/serializers.py:
```diff
@@ -4955,7 +4955,10 @@ def get_instances(self, obj):
     changes = serializers.SerializerMethodField()
-    object_association = serializers.SerializerMethodField()
+    object_association = serializers.SerializerMethodField(
+        help_text="When present, shows the field name of the role or relationship that changed.")
```
I believe this needs translation, `_(...)`

jbradberry committed:
i18n

AlexSCorey commented:
This looks good.  Thanks for doing the ui work @jbradberry.

ryanpetrello commented:
@dsesami you have some time to give this a try re: https://github.com/ansible/awx/issues/3841

dsesami commented:
going to take a look at this. will regate as well

dsesami commented:
regate
